src/nupic/algorithms/Connections.cpp

```diff
@@ -418,78 +389,35 @@ bool segmentIncreasingOrder(const SegmentOverlap& a, const SegmentOverlap& b)
   return a.segment < b.segment;
 }
 
-void Connections::computeActivity(const vector<Cell>& input,
+void Connections::computeActivity(const vector<CellIdx>& input,
                                   Permanence activePermanenceThreshold,
                                   SynapseIdx activeSynapseThreshold,
                                   Permanence matchingPermanenceThreshold,
                                   SynapseIdx matchingSynapseThreshold,
-                                  vector<SegmentOverlap>& outActiveSegments,
-                                  vector<SegmentOverlap>& outMatchingSegments,
-                                  bool recordIteration)
+                                  vector<SegmentOverlap>& activeSegmentsOut,
+                                  vector<SegmentOverlap>& matchingSegmentsOut)
+  const
 {
-  vector<UInt32> numActiveSynapsesForSegment(nextFlatIdx_, 0);
-  vector<UInt32> numMatchingSynapsesForSegment(nextFlatIdx_, 0);
+  SegmentExcitationTally excitations(*this, activePermanenceThreshold,
```

Comment:
The risk with this approach is the overhead for every call to `addActivePresynapticCell`. One common pattern is to take an iterator rather than explicit type (like vector). Then the caller can define an iterator class that operates on a vector, multiple vectors, or whatever else. This prevents possible misuse of the class and calling things out of order / minimizes the surface area to test.

What do you think?